File: examples/19-unix-sockets-fd-passing/go/main.go
// pmon v5 — a process supervisor with a UNIX-socket control plane.
//
// `pmon supervise` spawns a child command (its own process group,
// stdout+stderr appended to a log file), restarts it 300 ms after every
// exit, and serves a SOCK_STREAM control socket. `pmon pmctl` is the client
// side of the same binary:
//
//	status  -> "child=<pid> uptime=<s>s restarts=<n>"
//	stop    -> "stopping", then the supervisor tears down and exits 0
//	logfd   -> the supervisor passes its OPEN log file descriptor through an
//	           SCM_RIGHTS control message; pmctl reads the last 3 lines
//	           straight off the received fd ("via-fd: ..."), never touching
//	           the path.
//
// Where C++ and Rust build the cmsg by hand / via nix, Go rides
// net.UnixConn: WriteMsgUnix with unix.UnixRights on the way out,
// ReadMsgUnix + unix.ParseSocketControlMessage + unix.ParseUnixRights on the
// way in. The reaper is a goroutine; restart backoff and stop are a select.
package main

import (
	"errors"
	"fmt"
	"net"
	"os"
	"os/exec"
	"strings"
	"sync"
	"syscall"
	"time"

	"golang.org/x/sys/unix"
)

const (
	restartBackoff = 300 * time.Millisecond
	maxLogRead     = 1 << 20 // pmctl reads at most 1 MiB of log
)

func usage() int {
	fmt.Fprintln(os.Stderr, "usage: pmon supervise --ctl SOCK --log FILE -- CMD [ARG...]")
	fmt.Fprintln(os.Stderr, "       pmon pmctl --ctl SOCK <status|stop|logfd>")
	return 2
}

// ---------------------------------------------------------------------------
// Child lifecycle
// ---------------------------------------------------------------------------

type state struct {
	mu       sync.Mutex
	child    int // pid of the current child
	started  time.Time
	restarts int
	stopping bool
}

// spawnChild starts argv in its own process group with stdout+stderr going to
// the (O_APPEND) log file — the fd is inherited directly, no pipe goroutine.
func spawnChild(argv []string, logFile *os.File) (*exec.Cmd, error) {
	cmd := exec.Command(argv[0], argv[1:]...)
	cmd.Stdout = logFile
	cmd.Stderr = logFile
	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
	if err := cmd.Start(); err != nil {
		return nil, fmt.Errorf("exec %s: %w", argv[0], err)
	}
	return cmd, nil
}

func exitCodeOf(err error) int {
	var ee *exec.ExitError
	if errors.As(err, &ee) {
		if ws, ok := ee.Sys().(syscall.WaitStatus); ok {
			if ws.Signaled() {
				return 128 + int(ws.Signal())
			}
			return ws.ExitStatus()
		}
	}
	if err != nil {
		return 1
	}
	return 0
}

func logLine(logFile *os.File, line string) {
	// One write per line: O_APPEND makes it an atomic append.
	_, _ = logFile.WriteString(line + "\n")
}

// reaper waits for the current child; unless we are stopping, backs off
// 300 ms (or bails on stopCh) and respawns. Closes done when the final child
// is collected.
func reaper(s *state, cmd *exec.Cmd, argv []string, logFile *os.File, stopCh <-chan struct{}, done chan<- struct{}) {
	defer close(done)
	for {
		pid := cmd.Process.Pid
		err := cmd.Wait()
		s.mu.Lock()
		stopping := s.stopping
		s.mu.Unlock()
		if stopping {
			return
		}
		fmt.Fprintf(os.Stderr, "pmon: child pid=%d exited status=%d\n", pid, exitCodeOf(err))
		select {
		case <-stopCh:
			return
		case <-time.After(restartBackoff):
		}
		next, serr := spawnChild(argv, logFile)
		if serr != nil {
			fmt.Fprintf(os.Stderr, "pmon: error: %v\n", serr)
			return
		}
		s.mu.Lock()
		stopping = s.stopping
		var restartsNow int
		if !stopping {
			s.child = next.Process.Pid
			s.started = time.Now()
			s.restarts++
			restartsNow = s.restarts
		}
		s.mu.Unlock()
		if stopping { // stop raced with the respawn: undo it
			_ = unix.Kill(-next.Process.Pid, unix.SIGTERM)
			_ = next.Wait()
			return
		}
		fmt.Fprintf(os.Stderr, "pmon: restart %d child pid=%d\n", restartsNow, next.Process.Pid)
		logLine(logFile, fmt.Sprintf("pmon: start child pid=%d", next.Process.Pid))
		cmd = next
	}
}

// ---------------------------------------------------------------------------
// supervise
// ---------------------------------------------------------------------------

func peerCreds(conn *net.UnixConn) {
	raw, err := conn.SyscallConn()
	if err != nil {
		return
	}
	_ = raw.Control(func(fd uintptr) {
		cred, cerr := unix.GetsockoptUcred(int(fd), unix.SOL_SOCKET, unix.SO_PEERCRED)
		if cerr == nil {
			fmt.Fprintf(os.Stderr, "pmon: ctl connect uid=%d pid=%d\n", cred.Uid, cred.Pid)
		}
	})
}

func readCommand(conn *net.UnixConn) string {
	var cmd []byte
	one := make([]byte, 1)
	for len(cmd) < 64 {
		n, err := conn.Read(one)
		if n <= 0 || err != nil || one[0] == '\n' {
			break
		}
		cmd = append(cmd, one[0])
	}
	return string(cmd)
}

func cmdSupervise(ctl, logPath string, childArgv []string) error {
	// O_RDWR, not O_WRONLY: the fd handed out via SCM_RIGHTS shares this open
	// file description, and pmctl must be able to read from it.
	logFile, err := os.OpenFile(logPath, os.O_CREATE|os.O_RDWR|os.O_APPEND, 0o644)
	if err != nil {
		return fmt.Errorf("open %s: %w", logPath, err)
	}
	defer logFile.Close()

	_ = os.Remove(ctl) // stale socket from a previous run
	listener, err := net.ListenUnix("unix", &net.UnixAddr{Name: ctl, Net: "unix"})
	if err != nil {
		return fmt.Errorf("listen %s: %w", ctl, err)
	}
	// The stop path removes the socket explicitly; keep Go from doing it early.
	listener.SetUnlinkOnClose(false)
	defer listener.Close()
	fmt.Fprintf(os.Stderr, "pmon: listening on %s\n", ctl)

	first, err := spawnChild(childArgv, logFile)
	if err != nil {
		_ = os.Remove(ctl)
		return err
	}
	fmt.Fprintf(os.Stderr, "pmon: started child pid=%d\n", first.Process.Pid)
	logLine(logFile, fmt.Sprintf("pmon: start child pid=%d", first.Process.Pid))

	s := &state{child: first.Process.Pid, started: time.Now()}
	stopCh := make(chan struct{})
	done := make(chan struct{})
	go reaper(s, first, childArgv, logFile, stopCh, done)

	for {
		conn, aerr := listener.AcceptUnix()
		if aerr != nil {
			return fmt.Errorf("accept: %w", aerr)
		}
		peerCreds(conn)
		cmd := readCommand(conn)
		switch cmd {
		case "status":
			s.mu.Lock()
			reply := fmt.Sprintf("child=%d uptime=%ds restarts=%d\n",
				s.child, int(time.Since(s.started)/time.Second), s.restarts)
			s.mu.Unlock()
			_, _ = conn.Write([]byte(reply))
			conn.Close()
		case "logfd":
			rights := unix.UnixRights(int(logFile.Fd()))
			if _, _, werr := conn.WriteMsgUnix([]byte("ok\n"), rights, nil); werr != nil {
				fmt.Fprintf(os.Stderr, "pmon: sendmsg: %v\n", werr)
			}
			conn.Close()
		case "stop":
			_, _ = conn.Write([]byte("stopping\n"))
			conn.Close() // deliver the reply before tearing down
			fmt.Fprintln(os.Stderr, "pmon: stopping")
			s.mu.Lock()
			s.stopping = true
			child := s.child
			s.mu.Unlock()
			close(stopCh)
			if child > 0 {
				_ = unix.Kill(-child, unix.SIGTERM) // whole process group; ESRCH is fine
			}
			<-done
			_ = os.Remove(ctl)
			return nil
		default:
			_, _ = conn.Write([]byte("err unknown command\n"))
			conn.Close()
		}
	}
}

// ---------------------------------------------------------------------------
// pmctl
// ---------------------------------------------------------------------------

func recvLogFd(conn *net.UnixConn) (int, error) {
	buf := make([]byte, 16)
	oob := make([]byte, unix.CmsgSpace(4))
	_, oobn, _, _, err := conn.ReadMsgUnix(buf, oob)
	if err != nil {
		return -1, fmt.Errorf("recvmsg: %w", err)
	}
	msgs, err := unix.ParseSocketControlMessage(oob[:oobn])
	if err != nil {
		return -1, fmt.Errorf("parse cmsg: %w", err)
	}
	for _, m := range msgs {
		fds, ferr := unix.ParseUnixRights(&m)
		if ferr == nil && len(fds) > 0 {
			return fds[0], nil
		}
	}
	return -1, errors.New("no SCM_RIGHTS control message in reply")
}

func logTailViaFd(fd int, count int) ([]string, error) {
	var data []byte
	chunk := make([]byte, 4096)
	var off int64
	for len(data) < maxLogRead {
		n, err := unix.Pread(fd, chunk, off)
		if err != nil {
			return nil, fmt.Errorf("pread: %w", err)
		}
		if n == 0 {
			break
		}
		data = append(data, chunk[:n]...)
		off += int64(n)
	}
	text := strings.TrimSuffix(string(data), "\n")
	if text == "" {
		return nil, nil
	}
	lines := strings.Split(text, "\n")
	if len(lines) > count {
		lines = lines[len(lines)-count:]
	}
	return lines, nil
}

func cmdPmctl(ctl, action string) error {
	addr := &net.UnixAddr{Name: ctl, Net: "unix"}
	conn, err := net.DialUnix("unix", nil, addr)
	if err != nil {
		return fmt.Errorf("connect %s: %w", ctl, err)
	}
	defer conn.Close()
	if _, err := conn.Write([]byte(action + "\n")); err != nil {
		return fmt.Errorf("write %s: %w", ctl, err)
	}

	if action == "logfd" {
		fd, ferr := recvLogFd(conn)
		if ferr != nil {
			return ferr
		}
		defer unix.Close(fd)
		lines, terr := logTailViaFd(fd, 3)
		if terr != nil {
			return terr
		}
		for _, line := range lines {
			fmt.Printf("via-fd: %s\n", line)
		}
		return nil
	}

	// status / stop: the supervisor replies with one line and closes.
	var reply []byte
	buf := make([]byte, 256)
	for {
		n, rerr := conn.Read(buf)
		if n > 0 {
			reply = append(reply, buf[:n]...)
		}
		if rerr != nil {
			break
		}
	}
	text := strings.TrimRight(string(reply), "\n")
	if text == "" {
		return errors.New("empty reply from supervisor")
	}
	if strings.HasPrefix(text, "err ") {
		return errors.New(strings.TrimPrefix(text, "err "))
	}
	fmt.Println(text)
	return nil
}

// ---------------------------------------------------------------------------
// CLI
// ---------------------------------------------------------------------------

func main() {
	args := os.Args[1:]
	if len(args) == 0 {
		os.Exit(usage())
	}

	switch args[0] {
	case "supervise":
		var ctl, logPath string
		var child []string
		i := 1
		for i < len(args) {
			switch {
			case args[i] == "--ctl" && i+1 < len(args):
				ctl = args[i+1]
				i += 2
			case args[i] == "--log" && i+1 < len(args):
				logPath = args[i+1]
				i += 2
			case args[i] == "--":
				child = args[i+1:]
				i = len(args)
			default:
				os.Exit(usage())
			}
		}
		if ctl == "" || logPath == "" || len(child) == 0 {
			os.Exit(usage())
		}
		if err := cmdSupervise(ctl, logPath, child); err != nil {
			fmt.Fprintf(os.Stderr, "pmon: error: %v\n", err)
			os.Exit(1)
		}
	case "pmctl":
		var ctl, action string
		for i := 1; i < len(args); i++ {
			switch {
			case args[i] == "--ctl" && i+1 < len(args):
				ctl = args[i+1]
				i++
			case action == "":
				action = args[i]
			default:
				os.Exit(usage())
			}
		}
		if ctl == "" || (action != "status" && action != "stop" && action != "logfd") {
			os.Exit(usage())
		}
		if err := cmdPmctl(ctl, action); err != nil {
			fmt.Fprintf(os.Stderr, "pmctl: error: %v\n", err)
			os.Exit(1)
		}
	default:
		os.Exit(usage())
	}
}

File: examples/19-unix-sockets-fd-passing/rust/src/main.rs
// pmon v5 — a process supervisor with a UNIX-socket control plane.
//
// `pmon supervise` spawns a child command (its own process group,
// stdout+stderr appended to a log file), restarts it 300 ms after every
// exit, and serves a SOCK_STREAM control socket. `pmon pmctl` is the client
// side of the same binary:
//
//   status  -> "child=<pid> uptime=<s>s restarts=<n>"
//   stop    -> "stopping", then the supervisor tears down and exits 0
//   logfd   -> the supervisor passes its OPEN log file descriptor through an
//              SCM_RIGHTS control message; pmctl reads the last 3 lines
//              straight off the received fd ("via-fd: ..."), never touching
//              the path.
//
// The cmsg plumbing is nix::sys::socket sendmsg/recvmsg with
// ControlMessage::ScmRights — explicit, not a convenience crate — and the
// received fd is wrapped in an OwnedFd immediately so it closes on drop.

use std::fs::File;
use std::io::{IoSlice, IoSliceMut, Read, Write};
use std::os::fd::{AsRawFd, FromRawFd, OwnedFd, RawFd};
use std::os::unix::fs::FileExt;
use std::os::unix::net::{UnixListener, UnixStream};
use std::os::unix::process::{CommandExt, ExitStatusExt};
use std::process::{Child, Command, ExitCode, Stdio};
use std::sync::{Arc, Condvar, Mutex};
use std::time::{Duration, Instant};

use anyhow::{Context, Result, anyhow, bail};
use nix::sys::signal::{Signal, kill};
use nix::sys::socket::sockopt::PeerCredentials;
use nix::sys::socket::{
    ControlMessage, ControlMessageOwned, MsgFlags, getsockopt, recvmsg, sendmsg,
};
use nix::unistd::Pid;

const RESTART_BACKOFF: Duration = Duration::from_millis(300);
const MAX_LOG_READ: u64 = 1 << 20; // pmctl reads at most 1 MiB of log

fn usage() -> ExitCode {
    eprintln!("usage: pmon supervise --ctl SOCK --log FILE -- CMD [ARG...]");
    eprintln!("       pmon pmctl --ctl SOCK <status|stop|logfd>");
    ExitCode::from(2)
}

// ---------------------------------------------------------------------------
// Child lifecycle
// ---------------------------------------------------------------------------

struct State {
    child: i32, // pid of the current child
    started: Instant,
    restarts: u32,
    stopping: bool,
    child_done: bool, // reaper has collected the final child
}

type Shared = Arc<(Mutex<State>, Condvar)>;

/// Spawn argv in its own process group with stdout+stderr appended to the
/// log file — the fd is inherited directly, no pipe thread.
fn spawn_child(argv: &[String], log: &File) -> Result<Child> {
    let mut cmd = Command::new(&argv[0]);
    cmd.args(&argv[1..])
        .stdout(Stdio::from(log.try_clone().context("dup log fd")?))
        .stderr(Stdio::from(log.try_clone().context("dup log fd")?))
        .process_group(0); // own process group, so stop can signal the whole tree
    cmd.spawn().with_context(|| format!("exec {}", argv[0]))
}

fn exit_code_of(status: std::process::ExitStatus) -> i32 {
    if let Some(code) = status.code() {
        code
    } else {
        128 + status.signal().unwrap_or(1)
    }
}

/// One write per line: O_APPEND makes it an atomic append.
fn log_line(mut log: &File, line: &str) {
    let _ = log.write_all(format!("{line}\n").as_bytes());
}

fn mark_done(shared: &Shared) {
    let (lock, cv) = &**shared;
    lock.lock().unwrap().child_done = true;
    cv.notify_all();
}

/// Reaper thread: wait for the current child; unless we are stopping, back
/// off 300 ms and respawn. Ends by reporting child_done through the condvar.
fn reaper(shared: Shared, mut child: Child, argv: Vec<String>, log: File) {
    let (lock, cv) = &*shared;
    loop {
        let pid = child.id() as i32;
        let status = match child.wait() {
            Ok(status) => status,
            Err(err) => {
                eprintln!("pmon: error: waitpid: {err}");
                mark_done(&shared);
                return;
            }
        };
        if lock.lock().unwrap().stopping {
            mark_done(&shared);
            return;
        }
        eprintln!(
            "pmon: child pid={} exited status={}",
            pid,
            exit_code_of(status)
        );
        {
            let guard = lock.lock().unwrap();
            let (guard, _timed_out) = cv
                .wait_timeout_while(guard, RESTART_BACKOFF, |s| !s.stopping)
                .unwrap();
            if guard.stopping {
                drop(guard);
                mark_done(&shared);
                return;
            }
        }
        let next = match spawn_child(&argv, &log) {
            Ok(next) => next,
            Err(err) => {
                eprintln!("pmon: error: {err:#}");
                mark_done(&shared);
                return;
            }
        };
        let npid = next.id() as i32;
        let mut restarts_now = 0;
        let stop_now = {
            let mut s = lock.lock().unwrap();
            if !s.stopping {
                s.child = npid;
                s.started = Instant::now();
                s.restarts += 1;
                restarts_now = s.restarts;
            }
            s.stopping
        };
        if stop_now {
            // stop raced with the respawn: undo it
            let mut next = next;
            let _ = kill(Pid::from_raw(-npid), Signal::SIGTERM);
            let _ = next.wait();
            mark_done(&shared);
            return;
        }
        eprintln!("pmon: restart {restarts_now} child pid={npid}");
        log_line(&log, &format!("pmon: start child pid={npid}"));
        child = next;
    }
}

// ---------------------------------------------------------------------------
// supervise
// ---------------------------------------------------------------------------

fn read_command(conn: &mut UnixStream) -> String {
    let mut cmd = Vec::new();
    let mut one = [0u8; 1];
    while cmd.len() < 64 {
        match conn.read(&mut one) {
            Ok(1) if one[0] != b'\n' => cmd.push(one[0]),
            _ => break,
        }
    }
    String::from_utf8_lossy(&cmd).into_owned()
}

fn send_log_fd(conn: &UnixStream, log: &File) {
    let payload = [IoSlice::new(b"ok\n")];
    let fds = [log.as_raw_fd()];
    let cmsg = [ControlMessage::ScmRights(&fds)];
    if let Err(err) = sendmsg::<()>(conn.as_raw_fd(), &payload, &cmsg, MsgFlags::empty(), None) {
        eprintln!("pmon: sendmsg: {err}");
    }
}

fn cmd_supervise(ctl: &str, log_path: &str, child_argv: Vec<String>) -> Result<()> {
    // read(true), too: the fd handed out via SCM_RIGHTS shares this open
    // file description, and pmctl must be able to read from it.
    let log = File::options()
        .create(true)
        .read(true)
        .append(true)
        .open(log_path)
        .with_context(|| format!("open {log_path}"))?;

    let _ = std::fs::remove_file(ctl); // stale socket from a previous run
    let listener = UnixListener::bind(ctl).with_context(|| format!("listen {ctl}"))?;
    eprintln!("pmon: listening on {ctl}");

    let first = match spawn_child(&child_argv, &log) {
        Ok(first) => first,
        Err(err) => {
            let _ = std::fs::remove_file(ctl);
            return Err(err);
        }
    };
    let first_pid = first.id() as i32;
    eprintln!("pmon: started child pid={first_pid}");
    log_line(&log, &format!("pmon: start child pid={first_pid}"));

    let shared: Shared = Arc::new((
        Mutex::new(State {
            child: first_pid,
            started: Instant::now(),
            restarts: 0,
            stopping: false,
            child_done: false,
        }),
        Condvar::new(),
    ));
    let reaper_handle = {
        let shared = Arc::clone(&shared);
        let argv = child_argv.clone();
        let log = log.try_clone().context("dup log fd")?;
        std::thread::spawn(move || reaper(shared, first, argv, log))
    };

    for conn in listener.incoming() {
        let mut conn = conn.context("accept")?;

        if let Ok(cred) = getsockopt(&conn, PeerCredentials) {
            eprintln!("pmon: ctl connect uid={} pid={}", cred.uid(), cred.pid());
        }

        let cmd = read_command(&mut conn);
        let (lock, cv) = &*shared;
        match cmd.as_str() {
            "status" => {
                let reply = {
                    let s = lock.lock().unwrap();
                    format!(
                        "child={} uptime={}s restarts={}\n",
                        s.child,
                        s.started.elapsed().as_secs(),
                        s.restarts
                    )
                };
                let _ = conn.write_all(reply.as_bytes());
            }
            "logfd" => send_log_fd(&conn, &log),
            "stop" => {
                let _ = conn.write_all(b"stopping\n");
                drop(conn); // deliver the reply before tearing down
                eprintln!("pmon: stopping");
                let child = {
                    let mut s = lock.lock().unwrap();
                    s.stopping = true;
                    s.child
                };
                cv.notify_all();
                if child > 0 {
                    // whole process group; ESRCH is fine
                    let _ = kill(Pid::from_raw(-child), Signal::SIGTERM);
                }
                {
                    let guard = lock.lock().unwrap();
                    let _guard = cv.wait_while(guard, |s| !s.child_done).unwrap();
                }
                let _ = reaper_handle.join();
                let _ = std::fs::remove_file(ctl);
                return Ok(());
            }
            _ => {
                let _ = conn.write_all(b"err unknown command\n");
            }
        }
    }
    Ok(())
}

// ---------------------------------------------------------------------------
// pmctl
// ---------------------------------------------------------------------------

fn recv_log_fd(conn: &UnixStream) -> Result<OwnedFd> {
    let mut payload = [0u8; 16];
    let mut iov = [IoSliceMut::new(&mut payload)];
    let mut cmsg_buf = nix::cmsg_space!(RawFd);
    let msg = recvmsg::<()>(
        conn.as_raw_fd(),
        &mut iov,
        Some(&mut cmsg_buf),
        MsgFlags::MSG_CMSG_CLOEXEC,
    )
    .context("recvmsg")?;
    for cmsg in msg.cmsgs().context("parse cmsg")? {
        if let ControlMessageOwned::ScmRights(fds) = cmsg
            && let Some(&fd) = fds.first()
        {
            // SAFETY: the kernel installed this fd for us via SCM_RIGHTS;
            // we are its only owner.
            return Ok(unsafe { OwnedFd::from_raw_fd(fd) });
        }
    }
    bail!("no SCM_RIGHTS control message in reply")
}

fn log_tail_via_fd(fd: OwnedFd, count: usize) -> Result<Vec<String>> {
    let file = File::from(fd);
    let mut data = Vec::new();
    let mut chunk = [0u8; 4096];
    let mut off: u64 = 0;
    while off < MAX_LOG_READ {
        let n = file.read_at(&mut chunk, off).context("pread")?;
        if n == 0 {
            break;
        }
        data.extend_from_slice(&chunk[..n]);
        off += n as u64;
    }
    let text = String::from_utf8_lossy(&data);
    let text = text.strip_suffix('\n').unwrap_or(&text);
    if text.is_empty() {
        return Ok(Vec::new());
    }
    let lines: Vec<String> = text.split('\n').map(str::to_owned).collect();
    let skip = lines.len().saturating_sub(count);
    Ok(lines[skip..].to_vec())
}

fn cmd_pmctl(ctl: &str, action: &str) -> Result<()> {
    let mut conn = UnixStream::connect(ctl).with_context(|| format!("connect {ctl}"))?;
    conn.write_all(format!("{action}\n").as_bytes())
        .with_context(|| format!("write {ctl}"))?;

    if action == "logfd" {
        let fd = recv_log_fd(&conn)?;
        for line in log_tail_via_fd(fd, 3)? {
            println!("via-fd: {line}");
        }
        return Ok(());
    }

    // status / stop: the supervisor replies with one line and closes.
    let mut reply = String::new();
    conn.read_to_string(&mut reply).context("read reply")?;
    let text = reply.trim_end_matches('\n');
    if text.is_empty() {
        bail!("empty reply from supervisor");
    }
    if let Some(err) = text.strip_prefix("err ") {
        return Err(anyhow!("{err}"));
    }
    println!("{text}");
    Ok(())
}

// ---------------------------------------------------------------------------
// CLI
// ---------------------------------------------------------------------------

fn main() -> ExitCode {
    let args: Vec<String> = std::env::args().skip(1).collect();
    let Some(sub) = args.first() else {
        return usage();
    };

    match sub.as_str() {
        "supervise" => {
            let mut ctl = String::new();
            let mut log = String::new();
            let mut child: Vec<String> = Vec::new();
            let mut i = 1;
            while i < args.len() {
                if args[i] == "--ctl" && i + 1 < args.len() {
                    ctl = args[i + 1].clone();
                    i += 2;
                } else if args[i] == "--log" && i + 1 < args.len() {
                    log = args[i + 1].clone();
                    i += 2;
                } else if args[i] == "--" {
                    child = args[i + 1..].to_vec();
                    break;
                } else {
                    return usage();
                }
            }
            if ctl.is_empty() || log.is_empty() || child.is_empty() {
                return usage();
            }
            if let Err(err) = cmd_supervise(&ctl, &log, child) {
                eprintln!("pmon: error: {err:#}");
                return ExitCode::from(1);
            }
            ExitCode::SUCCESS
        }
        "pmctl" => {
            let mut ctl = String::new();
            let mut action = String::new();
            let mut i = 1;
            while i < args.len() {
                if args[i] == "--ctl" && i + 1 < args.len() {
                    ctl = args[i + 1].clone();
                    i += 2;
                } else if action.is_empty() {
                    action = args[i].clone();
                    i += 1;
                } else {
                    return usage();
                }
            }
            if ctl.is_empty() || !matches!(action.as_str(), "status" | "stop" | "logfd") {
                return usage();
            }
            if let Err(err) = cmd_pmctl(&ctl, &action) {
                eprintln!("pmctl: error: {err:#}");
                return ExitCode::from(1);
            }
            ExitCode::SUCCESS
        }
        _ => usage(),
    }
}
